Userid revocations and designated key revokers

zapu opened the pull request:
work in progress

Revoked identities won't show up in `Identities` map, but they can be accessed through `BadIdentities`. Also there is a `Revocation` field in `Identity` struct now.

Designated revokes end up in `UnverifiedRevocations`. Only revocations issued by designated keys are considered. Designation has to be signed and this is verified, but the consumer has to verify actual revocation (using the designee's key).

zapu committed:
Add userid revocations

zapu committed:
Add support for direct sigs with revocation keys

i.e. "designated revokers".

zapu committed:
go fmt

zapu committed:
Add openpgp.CheckDesignatedRevokers

It will go through entity.UnverifiedRevocations and try to verify them,
using passed KeyRing. If KeyRing is able to provide revocation issuers'
keys, if the signature matches, entity can be confirmed to be revoked by
trusted 3rd party.

zapu commented on openpgp/read.go:
```diff
@@ -461,3 +461,31 @@ func checkArmoredDetachedSignature(keyring KeyRing, signed, signature io.Reader)
+// issuer's key should be found in keyring, otherwise it will return
+// false, nil if keys are not found (meaning: revocation not
+// confirmed).
+func CheckDesignatedRevokers(keyring KeyRing, entity *Entity) (bool, *Key) {
```
Still unsure about the API here. Entity can be basically in 3 states now, in regards to revocations:
1) Not revoked
2) Maybe revoked (if there are 3rd party revocations we couldn't verify)
3) Definitely revoked (if there are 1st or 3rd party revocations that were verified)

maxtaco replied:
Where is this called from?

zapu replied:
Right now I'm calling this from client. So for example, `CheckDetachedSignature` would return a `signer` and then client would call `CheckDesignatedRevokers` with that signer. I'm not sure if it should be called automatically from e.g. `ReadMessage`.

maxtaco replied:
I see

maxtaco commented on openpgp/packet/public_key.go:
```diff
@@ -598,7 +598,7 @@ func (pk *PublicKey) CanSign() bool {
 }
 
-// VerifySignature returns nil iff sig is a valid signature, made by this
+// VerifySignature returns nil if sig is a valid signature, made by this
```
why these changes? `iff` == "if and only if".  I think it's intentional.

zapu replied:
I had no idea. I'll change these back. I run my editor with spell-checker and I occasionally correct a comment if it's near code I'm working on.

maxtaco reviewed (commented):
Some comments / questions.

zapu committed:
rollback "if and only if"

zapu commented on openpgp/keys.go:
```diff
@@ -500,13 +520,26 @@ EachPacket:
+
+					// Remove current from Identities, because the revocation signature
+					// might come after the certification signature.
+					delete(e.Identities, current.Name)
```
but few lines before is:
```
// NOTE! We might later see a revocation for this very same UID, and it
// won't be undone. We've preserved this feature from the original
// Google OpenPGP we forked from.
```

what should we do?

maxtaco replied:
If we make PGP more restrictive (i.e., disallow keys we previously allowed), we might be creating a ton of ~~bugs~~ support requests for ourselves down the road....

maxtaco replied:
The other thing is that we're not resetting `current.SelfSignature` in this case, which might be required.

maxtaco replied:
I'm OK to give this a try, but we have to be willing to back it out.

zapu replied:
Good point about `SelfSignature`. But for some keys it would have to go back and find another `SelfSignature`, from other uid.. So maybe lets not invalidate any uids, keep everything in `Identities` list but have the `Revocation` field in Identity is some code wanted to filter out revoked ones?

maxtaco replied:
Meaning get rid of `BadIdentities`?

zapu replied:
Yes. Otherwise whether revoked uid lands in Identities or BadIdentities is decided by order of signatures. If revocation comes first, it'll go to BadIdentities, if certification comes first, it'll go to Identities (but both cases will set `current.Revocation` field).

maxtaco replied:
And just to double-check, we previously weren't checking these revocations at all?

zapu replied:
Correct.

maxtaco replied:
ok, i'm on board.  nix the `BadIdentities`, sorry it was the wrong idea.

zapu committed:
Removed BadIdentities, all uids go to Identities

zapu committed:
Move CheckDesignatedRevokers to keys.go

zapu committed:
Change revoker function to FindVerifiedDesignatedRevoke

It returns (Signature, Key) instead of (bool, Key) if verifiable
signature is found.

maxtaco commented on openpgp/keys.go:
```diff
@@ -442,6 +452,13 @@ EachPacket:
+				// SigTypeIdentityRevocation). Handle these first,
+				// because key might have revocation coming from
+				// another key (designated revoke).
+				revocations = append(revocations, pkt)
```
What are implications of now allowing this case to fallthrough below?

maxtaco replied:
I.e., we're also appending this onto `current.Signatures`, whereas previous we weren't? Do I read that properly and is that OK?

zapu replied:
This is moved from line 503, i.e. it used to only append to `revocations` when `pkt.IssuerKeyId != nil && *pkt.IssuerKeyId != e.PrimaryKey.KeyId` did not evaluate to true. Now it always appends to `revocations` but that check is now at line 581 (for 1st party revocation) and there is additional logic regarding 3rd party revocations.

maxtaco replied:
That's not my question.  My question is that now these packets are going to hit line 551 in the new code, whereas previous it did not (since it was `if .. else` gated and handled by a different condition).  So all of these packets will be in the `current.Signatures` array as a result of line 551.  Does that matter?

zapu replied:
You're right, I missed that. It should not go to `current.Signatures`.

maxtaco replied:
ok, so if it's not intentional, let's fix it...

zapu replied:
Fixed one line below that (`continue`)

zapu committed:
Add more key revocation tests, fix ReadEntity

ReadEntity would add misplaced key revocation packet to
Identity.Signatures, in ReadEntity While these packets should not appear
like that in serialized keys, it was an unexpected behavior change and
was fixed. A new test was added that specifically checks for this
behavior with specially crafted key.
